src/python/pants/backend/java/package/fat_jar.py

```diff
@@ -0,0 +1,241 @@
+
+
+@rule
+async def package_fat_jar(
+    bash: BashBinary,
+    field_set: FatJarFieldSet,
+) -> BuiltPackage:
+    """
+    Constructs a "fat" JAR file (currently from Java sources only) by
+    1. compiling all Java sources
+    2. building a JAR containing all of those built class files
+    3. creating a fat jar with a broken ZIP index by concatenating all dependency JARs together,
+       followed by the thin JAR we created
+    4. using the unix `zip` utility's repair function to fix the broken fat jar
+    """
+
+    if field_set.root_address.value is None:
+        raise Exception("Needs a `root_address` argument")
+
+    if field_set.main_class.value is None:
+        raise Exception("Needs a `main` argument")
+
+    root_class_address = await Get(
+        Address,
+        AddressInput,
+        AddressInput.parse(field_set.root_address.value, relative_to=field_set.address.spec_path),
+    )
+
+    #
+    # 1. Collect Java source files and compile them all
+    #
+
+    transitive_targets = await Get(
+        TransitiveTargets, TransitiveTargetsRequest([root_class_address])
+    )
+
+    sources_to_compile = await Get(
+        CoarsenedTargets,
+        Addresses(
+            tgt.address for tgt in transitive_targets.closure if tgt.has_field(JavaSourceField)
+        ),
+    )
+
+    javac_request_gets = [
+        Get(FallibleCompiledClassfiles, CompileJavaSourceRequest(tgt)) for tgt in sources_to_compile
+    ]
+    fallible_class_files = await MultiGet(javac_request_gets)
+
+    failed = [i for i in fallible_class_files if i.result == CompileResult.FAILED]
+    if failed:
+        raise Exception(failed[0].stderr)
+
+    # TODO: better handle `i.output is None` case
+    class_files = [i.output for i in fallible_class_files if i.output is not None]
+    compiled_class_files_digest = await Get(Digest, MergeDigests(i.digest for i in class_files))
+    compiled_class_files_snapshot = await Get(Snapshot, Digest, compiled_class_files_digest)
+
+    #
+    # # 2. Produce thin JAR
+    #
+
+    main_class = field_set.main_class.value
+
+    manifest_content = FileContent(
+        "META-INF/MANIFEST.MF",
+        # NB(chrisjrn): we're joining strings with newlines, becuase the JAR manfiest format
+        # needs precise indentation, and _cannot_ start with a blank line. `dedent` seriously
+        # messes up those requirements.
+        "\n".join(
+            [
+                "Manifest-Version: 1.0",
+                f"Main-Class: {main_class}",
+                "",  # THIS BLANK LINE WILL BREAK EVERYTHING IF DELETED. DON'T DELETE IT.
+            ]
+        ).encode("utf-8"),
+    )
+
+    manifest_digest = await Get(Digest, CreateDigest([manifest_content]))
+    thin_jar_inputs = await Get(
+        Digest, MergeDigests([compiled_class_files_digest, manifest_digest])
+    )
+
+    thin_jar_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[
+                "/usr/bin/zip",
```

Comment:
You'll want to get the path to `zip` by doing a `BinaryPathRequest`. Look at how `BashBinary` gets its path.